File: src/julia/partition/algo/sgl_stc.jl
module Sgl_stc

include("../partition.jl")
using .Partition
using Base.Threads

function sgl_stc(arr::Vector{Vector{Int64}})
	result = Vector{Bool}(undef, length(arr))

	for i in 1:length(arr)
		result[i] = false
		size = length(arr[i])

		numOfCombinations = 1 << (size - 1)
		allNumbersMask = (1 << size) - 1

		problem_sum = partition_sum(arr[i], size, allNumbersMask)
		if problem_sum % 2 != 0
			continue
		end
		half_problem_sum = problem_sum ÷ 2

		found = Threads.Atomic{Bool}(false)
		Threads.@threads :static for j in 1:numOfCombinations
			if found[]
				continue
			end

			sum = partition_sum(arr[i], size, j)
			if sum == half_problem_sum
				atomic_cas!(found, false, true)
				continue
			end
		end
		result[i] = found[]

	end

	return result
end

end # module

File: src/go/partition/algo/sgl_stc.go
package algo

import (
	partition "golang/partition/lib"
	"runtime"
	"sync"
	"sync/atomic"
)

func Sgl_stc(arr [][]int) []int32 {
	var result []int32 = make([]int32, len(arr))
	nThreads := runtime.GOMAXPROCS(0) * 2 // Use more threads for CPU-bound work

	for i := 0; i < len(arr); i++ {
		result[i] = 0

		numOfCombinations := int64(1) << (len(arr[i]) - 1)
		allNumbersMask := (int64(1) << len(arr[i])) - 1

		problemSum := partition.Partition_sum(arr[i], allNumbersMask)
		if problemSum%2 != 0 {
			continue
		}
		halfSum := problemSum / 2
		chunkSize := numOfCombinations / int64(nThreads)

		var wg sync.WaitGroup
		var found int32 // atomic flag instead of channel

		for w := 0; w < nThreads; w++ {
			start := int64(w) * chunkSize
			end := start + chunkSize
			if w == nThreads-1 {
				end = numOfCombinations
			}

			wg.Add(1)
			go func(start, end int64) {
				defer wg.Done()

				for j := start; j < end; j++ {
					if atomic.LoadInt32(&found) == 1 {
						return
					}

					if partition.Partition_sum(arr[i], j) == halfSum {
						atomic.StoreInt32(&found, 1)
						return
					}
				}
			}(start, end)
		}

		wg.Wait()
		result[i] = atomic.LoadInt32(&found)
	}

	return result
}
